Modules/unicodedata.c

```diff
@@ -1253,6 +1274,110 @@ unicodedata_UCD_lookup_impl(PyObject *self, const char *name,
     return PyUnicode_FromOrdinal(code);
 }
 
+typedef struct {
+    PyObject_HEAD
+    PyObject* str;
+    int pos;
+} GraphemeClusterIterator;
+
+void GCI_Del(PyObject* x)
+{
+    GraphemeClusterIterator *i = (GraphemeClusterIterator *)x;
+    Py_DECREF(i->str);
+    PyObject_Del(x);
+}
+
+static PyObject* GCI_iter(PyObject *self)
+{
+    Py_INCREF(self);
+    return self;
+}
+
+#include "grapheme_cluster_break_automaton.h"
+
+PyObject* GCI_iternext(PyObject *self)
+{
+    GraphemeClusterIterator *p = (GraphemeClusterIterator *)self;
+    int kind = PyUnicode_KIND(p->str);
+    void *pstr = PyUnicode_DATA(p->str);
+    if (PyUnicode_READ(kind, pstr, p->pos)) {
+        int start = p->pos;
+        GCBState s = STATE_sot;
+        while (1) {
+            if (!PyUnicode_READ(kind, pstr, p->pos)) {
+                return PyUnicode_Substring(p->str, start, p->pos);
+            }
+            Py_UCS4 chr = PyUnicode_READ(kind, pstr, p->pos);
+            int prop = _getrecord_ex(chr)->grapheme_cluster_break;
+            s = GRAPH_CLUSTER_AUTOMATON[s][prop];
+            if (s == STATE_BREAK) {
+                return PyUnicode_Substring(p->str, start, p->pos);
+            }
+            ++p->pos;
+        }
+    } else {
+        return NULL;
+    }
+}
+
+static PyTypeObject GraphemeClusterIteratorType = {
+    PyVarObject_HEAD_INIT(NULL, 0)
+    "unicodedata.GraphemeClusterIterator", /*tp_name*/
+    sizeof(GraphemeClusterIterator),       /*tp_basicsize*/
+    0,                         /*tp_itemsize*/
+    (destructor)GCI_Del,       /*tp_dealloc*/
+    0,                         /*tp_print*/
+    0,                         /*tp_getattr*/
+    0,                         /*tp_setattr*/
+    0,                         /*tp_compare*/
+    0,                         /*tp_repr*/
+    0,                         /*tp_as_number*/
+    0,                         /*tp_as_sequence*/
+    0,                         /*tp_as_mapping*/
+    0,                         /*tp_hash */
+    0,                         /*tp_call*/
+    0,                         /*tp_str*/
+    0,                         /*tp_getattro*/
+    0,                         /*tp_setattro*/
+    0,                         /*tp_as_buffer*/
+    Py_TPFLAGS_DEFAULT,
+    "Internal grapheme cluster iterator object.",           /* tp_doc */
+    0,  /* tp_traverse */
+    0,  /* tp_clear */
+    0,  /* tp_richcompare */
+    0,  /* tp_weaklistoffset */
+    GCI_iter,  /* tp_iter: __iter__() method */
+    GCI_iternext  /* tp_iternext: next() method */
+};
+
+/*[clinic input]
+unicodedata.UCD.break_graphemes
```

Comment:
I think it would be useful to support iterating a substring. Add optional `start` and `end` parameters as in `str.index()` or `re.Pattern.finditer()`.